saleor/product/utils/variant_prices.py

```diff
@@ -0,0 +1,47 @@
+from prices.taxed_money import TaxedMoney
+
+from ...discount.utils import fetch_active_discounts
+from ..models import Product
+
+
+def _get_product_minimal_variant_price(product, discounts):
+    # Start with the product's price as the minimal one
+    minimal_variant_price = product.price
+    for variant in product.variants.all():
+        variant_price = variant.get_price(discounts=discounts)
+        if isinstance(variant_price, TaxedMoney):
+            # The "variant.get_price" method may return "TaxedMoney" instance but
+            # because we haven't provided it with "taxes" kwarg we can ignore it
+            # and "cast" it to just "Money" by taking net/gross (we take gross
+            # just in case)
+            variant_price = variant_price.gross
+        minimal_variant_price = min(minimal_variant_price, variant_price)
+    return minimal_variant_price
+
+
+def update_product_minimal_variant_price(product, discounts=None):
+    if discounts is None:
+        discounts = fetch_active_discounts()
+    minimal_variant_price = _get_product_minimal_variant_price(product, discounts)
+    if product.minimal_variant_price != minimal_variant_price:
+        product.minimal_variant_price = minimal_variant_price
+        product.save(update_fields=["minimal_variant_price"])
+    return product
+
+
+def update_products_minimal_variant_prices(products=None, discounts=None):
+    if products is None:
+        products = Product.objects.prefetch_related("variants").iterator()
```

Comment:
it won't work.
**Only prefetch**
```
Product.objects.prefetch_related('variants')[0].variants.all()
DEBUG django.db.backends (0.001) SELECT "product_product"."id", "product_product"."publication_date", "product_product"."is_published", "product_product"."seo_title", "product_product"."seo_description", "product_product"."product_type_id", "product_product"."name", "product_product"."description", "product_product"."description_json", "product_product"."category_id", "product_product"."price", "product_product"."minimal_variant_price", "product_product"."attributes", "product_product"."updated_at", "product_product"."charge_taxes", "product_product"."weight", "product_product"."meta" FROM "product_product" ORDER BY "product_product"."name" ASC  LIMIT 1; args=() [PID:4554:MainThread]
DEBUG django.db.backends (0.001) SELECT "product_product"."id", "product_product"."publication_date", "product_product"."is_published", "product_product"."seo_title", "product_product"."seo_description", "product_product"."product_type_id", "product_product"."name", "product_product"."description", "product_product"."description_json", "product_product"."category_id", "product_product"."price", "product_product"."minimal_variant_price", "product_product"."attributes", "product_product"."updated_at", "product_product"."charge_taxes", "product_product"."weight", "product_product"."meta" FROM "product_product" ORDER BY "product_product"."name" ASC  LIMIT 1; args=() [PID:4554:MainThread]
DEBUG django.db.backends (0.001) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" IN (72); args=(72,) [PID:4554:MainThread]
DEBUG django.db.backends (0.001) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" IN (72); args=(72,) [PID:4554:MainThread]
Out[5]: <ProductVariantQueryset [<ProductVariant: 2l>, <ProductVariant: 500ml>, <ProductVariant: 1l>]>
next(Product.objects.prefetch_related('variants').iterator()).variants.all()
```

**Prefetch +iterator**
```
next(Product.objects.prefetch_related('variants').iterator()).variants.all()
DEBUG django.db.backends (0.001) DECLARE "_django_curs_4687144384_1" NO SCROLL CURSOR WITH HOLD FOR SELECT "product_product"."id", "product_product"."publication_date", "product_product"."is_published", "product_product"."seo_title", "product_product"."seo_description", "product_product"."product_type_id", "product_product"."name", "product_product"."description", "product_product"."description_json", "product_product"."category_id", "product_product"."price", "product_product"."minimal_variant_price", "product_product"."attributes", "product_product"."updated_at", "product_product"."charge_taxes", "product_product"."weight", "product_product"."meta" FROM "product_product" ORDER BY "product_product"."name" ASC; args=() [PID:4554:MainThread]
DEBUG django.db.backends (0.001) DECLARE "_django_curs_4687144384_1" NO SCROLL CURSOR WITH HOLD FOR SELECT "product_product"."id", "product_product"."publication_date", "product_product"."is_published", "product_product"."seo_title", "product_product"."seo_description", "product_product"."product_type_id", "product_product"."name", "product_product"."description", "product_product"."description_json", "product_product"."category_id", "product_product"."price", "product_product"."minimal_variant_price", "product_product"."attributes", "product_product"."updated_at", "product_product"."charge_taxes", "product_product"."weight", "product_product"."meta" FROM "product_product" ORDER BY "product_product"."name" ASC; args=() [PID:4554:MainThread]
Out[6]: DEBUG django.db.backends (0.001) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:MainThread]
DEBUG django.db.backends (0.001) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:MainThread]
<ProductVariantQueryset [<ProductVariant: 2l>, <ProductVariant: 500ml>, <ProductVariant: 1l>]>
DEBUG django.db.backends (0.002) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-59]
DEBUG django.db.backends (0.002) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-59]
DEBUG django.db.backends (0.005) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-56]
DEBUG django.db.backends (0.005) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-56]
DEBUG django.db.backends (0.002) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-62]
DEBUG django.db.backends (0.002) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-62]
DEBUG django.db.backends (0.002) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-66]
DEBUG django.db.backends (0.002) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-66]
```

Reply:
I've removed the `prefetch_related` code.